boards/apollo3/lora_things_plus/src/main.rs

```diff
@@ -139,10 +143,29 @@ struct LoRaThingsPlus {
     temperature: &'static capsules_extra::temperature::TemperatureSensor<'static>,
     humidity: &'static capsules_extra::humidity::HumiditySensor<'static>,
     air_quality: &'static capsules_extra::air_quality::AirQualitySensor<'static>,
+    rng: &'static capsules_core::rng::RngDriver<'static>,
     scheduler: &'static RoundRobinSched<'static>,
     systick: cortexm4::systick::SysTick,
 }
 
+fn atecc508a_wakeup() {
```

Comment:
This wakeup function is a little odd. We need to wake the device up before each request and this is the best way I can think of doing it.

Basically the board provides a wakeup function to the capsule. The capsule then calls that function before accessing the device.

It unfortunately contains `unsafe` and will probably break the I2C mux as well. The capsule can change the SDA line without going through the I2C capsule.

Any thoughts on better ways to do this?

Reply:
This sounds like a case for a [PeripheralManagement](https://github.com/tock/tock/blob/d2b44c36bc88780d2f7974db726f8027bdbe13ba/kernel/src/utilities/peripheral_management.rs) wrapper. Put the object to access the device inside of a PM struct, which can handle the wakeup (and poweroff if appropriate) correctly and consistently in one place; also encapsulates the `unsafe` if-needed.

The [sam4l i2c driver has a reasonably advanced example](https://github.com/tock/tock/blob/d2b44c36bc88780d2f7974db726f8027bdbe13ba/chips/sam4l/src/i2c.rs#L573) of using PM.

Why would it break the I2C mux? If it's a case of needing to 'own' the bus before starting transaction, I think you can use the `before/after_peripheral_management` hooks, possibly with a deferred call in the `before` path?

Reply:
> This sounds like a case for a [PeripheralManagement](https://github.com/tock/tock/blob/d2b44c36bc88780d2f7974db726f8027bdbe13ba/kernel/src/utilities/peripheral_management.rs) wrapper. Put the object to access the device inside of a PM struct, which can handle the wakeup (and poweroff if appropriate) correctly and consistently in one place; also encapsulates the `unsafe` if-needed.

I'm not sure that fixes the issue. From what I can tell the `PeripheralManager` would wrap the I2C chip implementation that bangs on the MMIO registers. That means that every I2C device in Tock needs to be wrapped. In this case it's the actual I2C device (`I2CDevice` in Tock) that needs to have the I2C bus pulled low before opreations.

> Why would it break the I2C mux?

Currently one device could be performing an I2C operation yet this capsule can still pull the SDA line low.

The `before_peripheral_access()` looks like the right direction, but I need it to apply to specifically this one capsule when it tried to perform an I2C operation. So I really need a `before_peripheral_access()` callback supported in `I2CDevice()` that can be set in the board file

Reply:
So, I think the _right_ way to do this would be to implement it in the capsule by sending `0x00` for a sufficient amount of time. There are two ways I can imagine doing this:

1. The maximum practical frequency is going to be 1MHz, and at that rate, 8 bytes should do it (1MHz is one bit per microsecond, we need 60 microseconds, so 60 bits, or 7.5 bytes; please check my math).
2. If we had a way to query the frequency of the I2CDevice (we do for SPI), could calculate this dynamically.

I think using method 1 is sufficiently efficient _and_ simple to do the trick.

Neither requires any changing of pin assignments.

Reply:
The problem is that it doesn't work.

After you send the first byte you don't get an Ack (there is nothing to Ack you) so the rest of the transaction fails.

One option is to dynamically change the I2C Bus speed, but that's not really any better I don't think

Reply:
Oh I see.... That's highly distressing. Damn...